### ISSUE TITLE
Unable to `compile to file` anything with float16_t

### ISSUE THREAD

**Issue by averrez** [halide/Halide#4040]
Hi everyone!

The following sample code fails with the error: `Can't represent a float with this many bits in C: float16`. Regular JIT compilation works like a charm though.

```
    Var x, y;
    ImageParam input{Float(16), 2};
    Param<float16_t> mul{"mul"};
        
    Func output;
    output(x, y) = input(x, y) * mul;
        
    // Any other variant of compile_to_ fails with the same error.
    output.compile_to_static_library("testlib", {input, mul}, "testfunc");
```

I'm working on algorithm that runs on Metal only and it relies on half type. It needs to be compiled for iOS. Is there any workaround?

Thanks!

**Comment by abadams** [comment]
I am currently fixing up our float16 support. I'll add your testcase to my branch.

**Comment by shoaibkamil** [comment]
I have a simple fix for this, but it may not be compatible with @abadams's fixes for float16.  See the branch [here](https://github.com/halide/Halide/tree/metal_float16).

The issue is that C++ doesn't have a float16 type, so it's unclear how to declare a function that takes a float16 input.  This branch passes it in as a uint16_t value.

**Comment by steven-johnson** [comment]
IMHO we should add something like `halide_float16_t` and `halide_bfloat16_t` as types in HalideRuntime.h to give us concrete runtime types for this. (Maybe implemented as struct-wrapped uint16 to ensure they aren't accidentally interchangeable with ints.)

**Comment by steven-johnson** [comment]
Did this get addressed when the float/bfloat16 changes landed?

**Comment by shoaibkamil** [comment]
This was fixed by #3667.
